### DISCUSSION TITLE
Wire thickness and distance control

### DISCUSSION THREAD

**Discussion by Astrei** [bfieldtools/bfieldtools#55] [Ideas]
I have been using the package for quite some time now. In my opinion, the biggest drawback of the stream function approach is the inability to control the distance between the resulting isocontours (future wires). That makes some solutions impossible to implement in reality.
With a least square method, by playing with the Tikhonov regularisation parameter one can reduce the gradient value and increase distancing.  One can also change the number of contours. But I found both methods not very useful. Sometimes you have to sacrifice too many contours to make it work and discretization error grows dramatically.

I came up with such a naive solution. I create several hundred contours, and then delete those that do not satisfy the minimum distance conditions:
```python
for contour_num in range(len(loops.paths)):
    for comp_contour in range(contour_num,len(loops.paths)):
        if contour_num != comp_contour and contour_num not in deletion_ind and comp_contour not in deletion_ind:
            min_distance = np.min(np.min(np.array([ np.linalg.norm(loops.discrete[comp_contour] - p, axis=1) for p in loops.discrete[contour_num]])))
            if min_distance <= 0.018:
                print('Minimum distance ' + str(min_distance)  + ' found for contours ' + str(comp_contour) + ' and ' + str(contour_num))
                print('Deleting contour ' + str(comp_contour))
                deletion_ind.append(comp_contour)

for index in sorted(deletion_ind, reverse=True):
    del loops.discrete[index]
```
For my task, it works surprisingly well, but I feel that I am breaking some fundamental rules here.  After this filtering, there will be no equal spacing between the stream function values. But is it that important? What could be the disadvantages?

**Comment by rzetter** [comment]
Hi!

Indeed, controlling the exact wire spacing/positioning when contouring the stream function is a challenge, especially if there are some "jumps" in the geometry where small differences in the stream function value creates large differences in wire location. 

As you say, it feels wrong to create wires with unequal stream-function spacing, and naive solutions may work for one problem but not another. With your method, you're keeping the topography of the "stream function landscape", but altering the gradient such that it is never too steep.

One somewhat similar method that keeps the equal spacing, but still allows for tuning of wire positions, is to instead of choosing N contours, generate N*L contours and then choose every Lth of those contours. One can then loop over different combinations of N and L and choose the combination with the best performance according to some criteria (e.g. magnetic field error, minimum wire spacing, etc.).

```
contours = np.linspace(coil.s.vert.min(), coil.s.vert.max(), N*L+1)[1:-1][0::L] #Skip first and last values as they won't generate contours, then pick every Lth
windings= coil.s.discretize(N_contours=N, contours = contours) #Second parameter overrides
```

If you want to explicitly limit the stream function gradient (i.e. the current/wire density), that can be done in the optimization stage by setting an extra linear constraint. This is easily possible, but you'll have to modify the optimize_streamfunctions() wrapper function. If you explore the function and its calls, you'll see how the different constraints are assembled. 

You'll need to append the gradient matrix (bfieldtools.mesh_calculus.gradient_matrix) to the G constraint matrix, like so:
`
constraint_matrix = np.append(constraint_matrix, gradient_matrix(coil.mesh), axis=0)
`
and then set upper and lower limits for the gradient for each vertex in the mesh.
`
upper_bound_stack = np.append(upper_bound_stack, gradient_upper_bound, axis=0)
`
`
lower_bound_stack = np.append(lower_bound_stack, gradient_lower_bound, axis=0)
`

Some experimentation with the numerical bound values will be required. note that in the current function the constraint matrix is scaled by its largest singular value before optimization. NB! This method places bounds on the gradient values component by component, not on its magnitude.
